### ISSUE TITLE
MPU 0.62 tests hang on macOS

### ISSUE THREAD

**Issue by twogee** [danaj/Math-Prime-Util#22]
Looks like tests in `81-bignum.t` are hanging at test 126

`ok(!eval { miller_rabin_random(10007,-4); },   "MRR(10007,-4)");`

Other diagnostics is `BigInt 0.47/1.999811, lib: GMP.  MPU::GMP 0.44`

The older version of MPU passes all tests.

**Comment by danaj** [comment]
Thanks for the report.  It's really helpful knowing where it's failing, as the CPAN testers matrix doesn't show that.

I know why it's happening in theory, but cannot reproduce it on any platform or version of Perl I have access to, including MacOS.  Somehow the -4 is returning on your (and some other testers) systems as what looks like a non-bigint object.  On every platform I try I get the expected "negative number" evaluation, which causes it to croak as expected.

I am baffled why this is happening on tester's systems but nothing I can see.  All the gcc compiler farm machines run this fine.  My mac with 5 different versions of Perl runs fine.  My dev machine with 30 versions of Perl doesn't fail on any of them.  I do not understand how "-4" is not getting processed correctly.

Since the XS routine has no idea what to do with it, it goes to GMP or PP.  In this case the GMP routine has a UV argument, so it hangs as it tries to perform trillions of tests.  I will update Math::Prime::Util::GMP to use a signed second argument as well as do the logical number of tests shortcut.

I'll also change the code in this module to hopefully better capture this.  I say hopefully only because I'm still not sure *why* it is failing.

**Comment by danaj** [comment]
Math::Prime::Util::GMP v0.45 released.  This will croak on negative inputs as well as limit the number of tests performed if it is past the Monier/Rabin limit.  So (10007,-4) croaks as expected, and (10007, ~0) will finish almost instantly since it knows there is no point in doing more than 3/4 * 10007 tests.

Leaving open for now until it's worked around in this module as well.

**Comment by danaj** [comment]
version 0.63 released which should stop this regardless of GMP back end.  I'll close the issue when I see enough tester results.  I was not able to reproduce the issue on any platform I have access to, so can't be absolutely sure at this point.

Apologies for this, as this is a terrible thing for test boxes.

**Comment by danaj** [comment]
The latest test matrix still shows version 0.63 failing on many test boxes.  Unfortunately it will be 1+ weeks before I can see any logs.  I'm very frustrated as for the life of me I cannot get this to fail, and I don't see how this test can be failing any more.

**Comment by twogee** [comment]
Now, it's test 130

`is(miller_rabin_random(1009), 1, "MRR(1009) = 1");   # runs one random base`

```
t/81-bignum.t .. 
1..134
# BigInt 0.47/1.999811, lib: GMP.  MPU::GMP 0.45
ok 1 - 100000982717289000001 is prime
ok 2 - 100000982717289000001 is probably prime
ok 3 - 100170437171734071001 is prime
ok 4 - 100170437171734071001 is probably prime
ok 5 - 777777777777777777777767 is prime
ok 6 - 777777777777777777777767 is probably prime
ok 7 - 777777777777777777777787 is prime
ok 8 - 777777777777777777777787 is probably prime
ok 9 - 877777777777777777777753 is prime
ok 10 - 877777777777777777777753 is probably prime
ok 11 - 877777777777777777777871 is prime
ok 12 - 877777777777777777777871 is probably prime
ok 13 - 87777777777777777777777795577 is prime
ok 14 - 87777777777777777777777795577 is probably prime
ok 15 - 890745785790123461234805903467891234681243 is prime
ok 16 - 890745785790123461234805903467891234681243 is probably prime
ok 17 - 618970019642690137449562111 is prime
ok 18 - 618970019642690137449562111 is probably prime
ok 19 - 777777777777777777777777 is not prime
ok 20 - 777777777777777777777777 is not probably prime
ok 21 - 877777777777777777777777 is not prime
ok 22 - 877777777777777777777777 is not probably prime
ok 23 - 87777777777777777777777795475 is not prime
ok 24 - 87777777777777777777777795475 is not probably prime
ok 25 - 890745785790123461234805903467891234681234 is not prime
ok 26 - 890745785790123461234805903467891234681234 is not probably prime
ok 27 - 564132928021909221014087501701 is not prime
ok 28 - 564132928021909221014087501701 is not probably prime
ok 29 - 6003094289670105800312596501 is not prime
ok 30 - 6003094289670105800312596501 is not probably prime
ok 31 - 318665857834031151167461 is not prime
ok 32 - 318665857834031151167461 is not probably prime
ok 33 - 21652684502221 is not prime
ok 34 - 21652684502221 is not probably prime
ok 35 - 3825123056546413051 is not prime
ok 36 - 3825123056546413051 is not probably prime
ok 37 - 59276361075595573263446330101 is not prime
ok 38 - 59276361075595573263446330101 is not probably prime
ok 39 - 75792980677 is not prime
ok 40 - 75792980677 is not probably prime
ok 41 - 3317044064679887385961981 is not prime
ok 42 - 3317044064679887385961981 is not probably prime
ok 43 - 65635624165761929287 is prime
ok 44 - 65635624165761929287 is provably prime
ok 45 - 1162566711635022452267983 is prime
ok 46 - 1162566711635022452267983 is provably prime
ok 47 - 77123077103005189615466924501 is prime
ok 48 - 77123077103005189615466924501 is provably prime
ok 49 - 3991617775553178702574451996736229 is prime
ok 50 - 3991617775553178702574451996736229 is provably prime
ok 51 - 273952953553395851092382714516720001799 is prime
ok 52 - 273952953553395851092382714516720001799 is provably prime
ok 53 - primes( 2^66, 2^66 + 100 )
ok 54 - twin_primes( 18446744073709558000, +1000)
ok 55 - next_prime(777777777777777777777777)
ok 56 - prev_prime(777777777777777777777777)
ok 57 - iterator 3 primes starting at 10^24+910
ok 58 - prime_count(87..7752, 87..7872)
ok 59 - 564132928021909221014087501701 is a strong pseudoprime to bases 2,3,5,7,11,13,17,19,23,29,31,37,325,9375
ok 60 - 6003094289670105800312596501 is a strong pseudoprime to bases 2,3,5,7,11,13,17,19,23,29,31,37,61,325,9375
ok 61 - 318665857834031151167461 is a strong pseudoprime to bases 2,3,5,7,11,13,17,19,23,29,31,37,325,9375
ok 62 - 21652684502221 is a strong pseudoprime to bases 2,7,37,61,9375
ok 63 - 3825123056546413051 is a strong pseudoprime to bases 2,3,5,7,11,13,17,19,23,29,31,325,9375
ok 64 - 59276361075595573263446330101 is a strong pseudoprime to bases 2,3,5,7,11,13,17,19,23,29,31,37,325,9375
ok 65 - 75792980677 is a strong pseudoprime to bases 2
ok 66 - 3317044064679887385961981 is a strong pseudoprime to bases 2,3,5,7,11,13,17,19,23,29,31,37,73,325,9375
ok 67 - PC approx(31415926535897932384)
ok 68 - prime count bounds for 31415926535897932384 are in the right order
ok 69 - PC lower with RH
ok 70 - PC upper with RH
ok 71 - PC lower
ok 72 - PC upper
ok 73 - factor(1234567890)
ok 74 - factor_exp(1234567890)
ok 75 - factor(23489223467134234890234680)
ok 76 - factor_exp(23489223467134234890234680)
ok 77 - factor(190128090927491)
ok 78 - factor_exp(190128090927491)
ok 79 - divisors(23489223467134234890234680)
ok 80 - moebius(618970019642690137449562110)
ok 81 - euler_phi(618970019642690137449562110)
ok 82 - carmichael_lambda(618970019642690137449562110)
ok 83 - jordan_totient(5,2188536338969724335807)
ok 84 - jordan totient using divisor_sum and moebius
ok 85 - Divisor sum of 100!
ok 86 - Divisor count(103\#)
ok 87 - Divisor sum(103\#)
ok 88 - sigma_2(103\#)
ok 89 - znorder 1
ok 90 - znorder 2
ok 91 - kronecker(..., ...)
ok 92 - znprimroot(333822190384002421914469856494764513809)
ok 93 - znlog(b,g,p): find k where b^k = g mod p
ok 94 - liouville(a x b x c) = -1
ok 95 - liouville(a x b x c x d) = 1
ok 96 - gcd(a,b,c)
ok 97 - gcd(a,b)
ok 98 - gcd of two primes = 1
ok 99 - lcm(p1,p2)
ok 100 - lcm(p1,p1)
ok 101 - lcm(a,b,c,d,e)
ok 102 - gcdext(a,b)
ok 103 - chinese([26,17179869209],[17,34359738421] = 103079215280
ok 104 - ispower(18475335773296164196) == 0
ok 105 - ispower(150607571^14) == 14
ok 106 - -7 ^ i for 0 .. 31
ok 107 - correct root from is_power for -7^i for 0 .. 31
ok 108 - random range prime isn't too small
ok 109 - random range prime isn't too big
ok 110 - random range prime is prime
ok 111 - random 25-digit prime is not too small
ok 112 - random 25-digit prime is not too big
ok 113 - random 25-digit prime is just right
ok 114 - random 80-bit prime is not too small
ok 115 - random 80-bit prime is not too big
ok 116 - random 80-bit prime is just right
ok 117 - random 180-bit strong prime is not too small
ok 118 - random 180-bit strong prime is not too big
ok 119 - random 180-bit strong prime is just right
ok 120 - random 80-bit Maurer prime is not too small
ok 121 - random 80-bit Maurer prime is not too big
ok 122 - random 80-bit Maurer prime is just right
ok 123 - 80-bit prime passes Miller-Rabin with 20 random bases
ok 124 - 80-bit composite fails Miller-Rabin with 40 random bases
ok 125 - MRR(undef,4)
ok 126 - MRR(10007,-4)
ok 127 - MRR(n,0) = 1
ok 128 - MRR(61,17) = 1
ok 129 - MRR(62,17) = 0
Usage: Math::Prime::Util::GMP::miller_rabin_random(strn, nbases, seedstr= 0) at t/81-bignum.t line 381.
# Looks like your test exited with 2 just after 129.
Dubious, test returned 2 (wstat 512, 0x200)
Failed 5/134 subtests 

Test Summary Report
-------------------
t/81-bignum.t (Wstat: 512 Tests: 129 Failed: 0)
  Non-zero exit status: 2
  Parse errors: Bad plan.  You planned 134 tests but ran 129.
Files=1, Tests=129,  0 wallclock secs ( 0.05 usr  0.01 sys +  0.28 cusr  0.02 csys =  0.36 CPU)
Result: FAIL
```

**Comment by danaj** [comment]
Thanks!

Apparently all the inputs are being rejected by the XS code as not valid numbers / bigint objects.  I wonder if something is wrong with "use bigint" on those machines.

    perl -E 'use bigint try="GMP,Pari"; say ref(1);'

should return "Math::BigInt".

Regardless, I will make sure this test passes, and thank you again for letting me know what happened.

**Comment by twogee** [comment]
```
> perl -E 'use bigint try="GMP,Pari"; say ref(1);'
Can't modify constant item in scalar assignment at -e line 1, near ""GMP,Pari";"
BEGIN not safe after errors--compilation aborted at -e line 1.
> perl -Mbigint -e 'print "$bigint::VERSION\n"'
0.47
```

**Comment by danaj** [comment]
Whoops, arrow turned into an equal.

```
$ perl -E 'use bigint try=>"GMP,Pari"; say ref(1);'
Math::BigInt
```

It doesn't seem to be the particular version:

`t/81-bignum.t ................ 1/134 # BigInt 0.47/1.999811, lib: GMP.  MPU::GMP 0.44`

**Comment by twogee** [comment]
Sorry for not checking the manual 😊
```
> perl -E 'use bigint try=>"GMP,Pari"; say ref(1);'
Math::BigInt::Lite
```
0.47 is the version of `bignum`, 1.999811 is the version of `Math::BigInt`.

I just tried install MPU 0.63 on my other Mac and all the tests passed... I will try to investigate the differences between them; however, the check above returns `Math::BigInt::Lite` on both of them. Perhaps I'll just try to look at the output from `cpanp b`.

**Comment by twogee** [comment]
My other Mac lacked a handful of modules, but no obvious culprit (sorry about the wall of text)
```
< AnyData 0.12
< AnyEvent 7.13
< Archive::Any 0.0945
< Array::Compare 3.0.0
< Array::Iterator 0.11
< Async::Interrupt 1.21
< B::Hooks::OP::Check::EntersubForCV 0.09
< Cache::LRU 0.04
< Class::AutoClass 1.56
< Class::Date 1.1.15
< Class::Member 1.6
< Class::Virtual 0.08
< Code::TidyAll 0.58
< Config::Crontab 1.44
< Config::INI 0.025
< Config::IniFiles 2.94
< Contextual::Return 0.004014
< Convert::BER 1.32
< Convert::EBCDIC 0.06
< Coro 6.511
< Crypt::DH 0.07
< Crypt::SSLeay 0.72
< Crypt::SmbHash 0.12
< Curses 1.36
< DBIx::Simple 1.35
< DBIx::SystemCatalog 0.06
< DCOP 0.038
< Data::CompactDump 0.04
< Data::Dump::Streamer 2.40
< Devel::BeginLift 0.001003
< Devel::Confess 0.009004
< Devel::Declare::Parser 0.020
< Devel::SimpleTrace 0.08
< Devel::Size 0.80
< Devel::StringInfo 0.04
< Devel::Trace 0.12
< Digest::MD4 1.9
< EV 4.22
< Email::ARF 0.010
< Email::Delete 2.002
< Email::Reply 1.204
< Email::Sender 1.300031
< Email::Thread 0.712
< Email::Valid 1.202
< Event 1.26
< Exception::Class::TryCatch 1.13
< Exporter::Declare 0.114
< ExtUtils::AutoInstall 0.64
< Fennec::Lite 0.004
< File::MMagic 1.30
< File::Tail 1.3
< File::Tempdir 0.02
< File::Zglob 0.11
< Getopt::Lucid 1.08
< Getopt::Lucid::Exception 1.08
< Graph::Easy 0.76
< Graph::Easy::As_svg 0.26
< Guard 1.023
< HTML::Clean 0.8
< HTTP::MultiPartParser 0.02
< HTTP::Proxy 0.304
< Hash::AutoHash 1.17
< Hash::AutoHash::Args 1.18
< Hints 0.02
< IO::AIO 4.34
< IO::Any 0.09
< IO::Multiplex 1.16
< IO::SessionData 1.03
< IPC::ShareLite 0.17
< LWP::Online 1.08
< List::Compare 0.53
< Logo::svplus 0.01
< MIME::Lite 3.030
< Mail::Audit 2.228
< Mail::DomainKeys 1.0
< Mail::ListDetector 1.04
< Mail::POP3Client 2.19
< Mail::Thread 2.55
< Math::Algebra::Symbols 1.27
< Math::Currency 0.51
< Math::SymbolicX::BigNum 0.02
< Math::SymbolicX::Complex 1.01
< Math::SymbolicX::ParserExtensionFactory 3.02
< Meta::Builder 0.003
< MetaCPAN::API 0.50
< Mock::Quick 1.111
< MooseX::Clone 0.06
< Net::CIDR 0.18
< Net::Domain::TLD 1.75
< Net::SNMP v6.0.1
< Net::SSH 0.09
< Net::SSL 2.86
< Net::Server::Mail 0.23
< Net::Telnet 3.04
< Pithub 0.01033
< Pod::Plainer 1.04
< Pod::WikiDoc 0.20
< PostScript::Poster 0.02
< Print::Printcap 0.01
< Regexp::Parser 0.21
< SVG 2.66
< Set::Object 1.35
< Sort::Fields 0.90
< Specio::Library::Path::Tiny 0.04
< String::CRC32 1.5
< Test::Class::Most 0.08
< Test::Code::TidyAll 0.58
< Test::DatabaseRow 2.04
< Test::Exception::LessClever 0.009
< Test::Manifest 2.02
< Test::TinyMocker 0.03
< Text::Shellwords 1.08
< Tie::DBI 1.06
< Tie::Hash::MultiValue 1.02
< Tie::RDBM 0.73
< Tie::Simple 1.04
< Time::CTime 2011.0505
< Time::DaysInMonth 99.1117
< Time::JulianDay 2011.0505
< Time::ParseDate 2015.103
< Time::Timezone 2015.0925
< Time::y2038 20100403
< Tk::Cloth 2.03
< Tk::Date 0.44
< Tk::FireButton 1.05
< Tk::MListbox 1.11
< Tk::NumEntry 2.06
< Tk::NumEntryPlain 1.10
< Tk::TFrame 2.02
< Types::Path::Tiny 0.005
< Version::Requirements 0.101023
< WWW::Mechanize::Pluggable 1.12
< WeakRef 0.01
< XML::RSS 1.59
< XML::Stream 1.24
< latest 0.03

```

**Comment by danaj** [comment]
That's the culprit.  I don't get that output anywhere until I install Math::BigInt::Lite.  All the sudden bigint changes.  This is hidden in the manual page for that module:

> The pragmas bigrat, bignum and bigint will automatically use Math::BigInt::Lite whenever possible.

so installing that module has fundamentally changed the behavior of core modules.

At least now I understand why we're getting this.  A new version of that module came out on Mar 1, so that would be why some test boxes have this happening and some don't.  Interesting that the module isn't listed there.

The issue isn't that that module's bigints aren't working right, it is that I don't recognize them.  I could decide that anything like "Math::BigInt*" must be a bigint.

**Comment by twogee** [comment]
Math::BigInt::Lite used to be broken on macOS for quite some time... and still, I'm quite sure I had Math::BigInt::Lite on another box where 0.63 installed without a problem.

**Comment by danaj** [comment]
I suspect it's that this causes a whole new path to be tested, that works on some machines and not on others.  I test with XS completely disabled, which goes through a Perl front end, but this is slightly different, where XS is still on and tries to start but will bail.

I'm not sure if this is a good thing for test boxes or bad.

It's also making me wonder if I should not worry about the type of the object but instead just stringify it.  Perhaps time for a redo of my whole argument parsing system...

**Comment by danaj** [comment]
It looks like this should have been fixed in 0.65.  I've tested today with and without the Lite module and it looks like things pass.  However I no longer have a Mac to test with.
